### DISCUSSION TITLE
PDF / email template: List uploaded file names from within repeater field?

### DISCUSSION THREAD

**Discussion by mandrasch** [verbb/formie#1566] [Q&A] [answered]
Hi,

I'm using https://verbb.io/craft-plugins/formie/docs/template-guides/pdf-templates.

I have a repeater field (`myUploads`) which contains multiple upload fields (`uploadField1`, `uploadField2`).

How can I create a simple list with file names in the attached PDF template for notification?

I tried various chatgpt suggestions, but did not get it quite right by now. 😉 

```
{# Assuming you have a submission model or variable named 'submission' and a repeater field named 'myUploads' #}
{% set repeaterFieldHandle = 'myUploads' %} {# Replace with your actual repeater field handle #}

{# Get the repeater field value from the submission #}
{% set repeaterField = submission.getFieldValue(repeaterFieldHandle) %}

{# Check if the repeater field has rows #}
{% if repeaterField | length %}
    <ul>
        {% for row in repeaterField %}
            <li>
                {# Loop through the upload fields in the repeater row #}
                {% for uploadField in row %}
                    {# Check if the field is an upload field and if it has a file (check the 'url' property) #}
                    {% if uploadField is iterable and uploadField.url is not empty %}
                        {# Get the file associated with the upload field #}
                        {% set file = craft.assets.id(uploadField.value).one() %}
                        
                        {# Check if the file exists #}
                        {% if file %}
                            {# Output the file name #}
                            {{ file.filename }}
                        {% endif %}
                    {% endif %}
                {% endfor %}
            </li>
        {% endfor %}
    </ul>
{% else %}
    No uploads found.
{% endif %}
```

Thanks very much in advance!

Best regards,
Matthias

**Comment by engram-design** [comment, marked as answer]
I would suggest brushing up on how [rendering submission content](https://verbb.io/craft-plugins/formie/user-guides/the-complete-guide-to-rendering-submission-content) works as that'll guide you through most cases.

But for your exact use-case:

```twig
{% set rows = submission.getFieldValue('myRepeaterField').all() %}

{% if rows | length %}
    <ul>
        {% for row in rows %}
            <li>
                {% set assets = row.getFieldValue('myFileUploadField').all() %}

                {% for asset in assets %}
                    {{ asset.filename }}
                {% endfor %}
            </li>
        {% endfor %}
    </ul>
{% else %}
    No uploads found.
{% endif %}
```

Keep in mind that a Repeater field is much like a Matrix or Super Table field, and a File Upload field is an Assets field. You'll notice the use of `all()` for the repeater field value, as that will return a `NestedRowQuery` value, much like Matrix returns a `MatrixBlockQuery`.

**Reply by mandrasch** [reply]
Hi @engram-design, 

thanks very much! Works! 🎉 

I was not aware of https://verbb.io/craft-plugins/formie/user-guides/the-complete-guide-to-rendering-submission-content

Just followed the docs link for `Submission` from https://verbb.io/craft-plugins/formie/docs/template-guides/pdf-templates#available-template-variables to https://verbb.io/craft-plugins/formie/docs/developers/submission and thought this is all regarding documentation. 

(Btw: Also noticed that `form` is not available within https://verbb.io/craft-plugins/formie/docs/template-guides/pdf-templates#available-template-variables, at least in my tests) 

Best Regards,
Matthias

**Reply by engram-design** [reply]
Good call on the missing `form` variable, that's been added for the next release. You can fetch it via `submission.getForm()` in the meantime.
